src/ast/codegen_llvm.cpp

```diff
@@ -1275,14 +1275,13 @@ void CodegenLLVM::visit(Unop &unop)
       }
       case bpftrace::Parser::token::MUL:
       {
-        int size = type.size;
-        if (type.IsPtrTy())
-        {
-          // When dereferencing a 32-bit integer, only read in 32-bits, etc.
-          size = type.GetPointeeTy()->size;
-        }
+        // When dereferencing a 32-bit integer, only read in 32-bits, etc.
+        int size = type.IsPtrTy() ? type.GetPointeeTy()->GetSize()
+                                  : type.GetSize();
+        auto as = type.IsPtrTy() ? type.GetAS() : AddrSpace::kernel;
```

Comment:
Yes this should remain backwards compatibility with how integer dereference works currently. As the old probe_read helpers are deprecated we shouldn't rely on them anymore, so we eneed to set the AS here.

Reply:
I think, this will resolve to probe kernel then, which is not correct. uprobe:./testprogs/intptrcast:fn { $a=*(reg("r15")+160);

Reply:
I don't think that differs from current behaviour where its also broken. But would be good to discuss in a separate issue

Reply:
If we want to preserve backward compatibility, then how about using Addrspace::none and bpf_probe_read()?

Reply:
hmm. Took a look at what the kernel does now, for systems that support overlapping addrspace it does fall back to _user if needed

```
#ifdef CONFIG_ARCH_HAS_NON_OVERLAPPING_ADDRESS_SPACE
BPF_CALL_3(bpf_probe_read_compat, void *, dst, u32, size,
           const void *, unsafe_ptr)
{
        if ((unsigned long)unsafe_ptr < TASK_SIZE) {
                return bpf_probe_read_user_common(dst, size,
                                (__force void __user *)unsafe_ptr);
        }
        return bpf_probe_read_kernel_common(dst, size, unsafe_ptr);
}
```

But on other architectures the whole fn is dropped.

https://github.com/torvalds/linux/commit/0ebeea8ca8a4d1d453ad299aef0507dab04f6e8d


Maybe this should keep addrspace::none for x86 and non-overlapping arch and use kernel for those that are overlapping? Than we can add ::user detection in later.

Reply:
I think, this is better:
* Overlapping addrspace architecture like x86:
Backward compatibility : AddrSpace::none

* For overlapping addrspace architecture, using Addrspace::kernel for userspace will not work.
So I would keep it as AddrSpace::none. 
* I would expect distribution to provide bpf_probe_read_{user/kernel} for this purpose.
* If the bpf_probe_read is present in older kernels, it will be used anyways.

Reply:
updated

Reply:
oh, I didn't know bpf_probe_read was dropped in some architectures...  If addrspace is none, then bpftrace emits a warning.  So, how about keeping addrspace::none in this case and use bpf_probe_read_kernel() if bpf_probe_read() is not available. A user can use uptr() to change the behavior if needed. And maybe we can add a better auto-detect mechanism in another PR.

Reply:
maybe we need to change this depending on whether arch has overlapping addrspace?
https://github.com/iovisor/bpftrace/blob/e6bbb9a925e405c1ec87790490d81227ac122032/src/ast/irbuilderbpf.cpp#L74-L79

Reply:
As this PR has some massive refactoring changes and a printf fix I don't want to block this on this discussion/change right now so I've restored the original AS behaviour. For non overlapping arch we should probably get autodetection working properly but that is too much for this PR. We can continue the discussion in #1507 :)

Reply:
ok